### PR TITLE
Adds esc_xml() and esc_xml__() functions.

### PR DESCRIPTION
**Pull request by pbiron**
### Description
These functions were originally introduced in #163.  This simply splits them out into their own PR.

This also introduces a new filter: `esc_xml`.

Both new functions and the new filter are the direct equivlants of `esc_html()`, `esc_html__()` and `esc_html` in core.

The only difference here is that HTML character entities are converted to their Unicode codepoints.

These new functions do **not** 100% guarantee that the resulting text is "safe" for use in an XML instance, because there are certain Unicode codepoints that are forbidding in XML see https://www.w3.org/TR/REC-xml/#charsets (note: the comment about the `Char` production in the XML spec is not completely accurate: many control characters are also not allowed) and it is an open question what to do when the text passed to these functions/filter contain characters which are not allowed in XML.

### Type of change
Please select the relevant options:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Enhancement (change which improves an existing feature. E.g., performance improvement, docs update, etc.)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Steps to test
Still needs unit tests

### Acceptance criteria
- [X] My code follows WordPress coding standards.
- [X] I have performed a self-review of my own code.
- [ ] If the changes are visual, I have cross browser / device tested.
- [X] I have commented my code, particularly in hard-to-understand areas.
- [X] My changes generate no new warnings.
- [ ] I have added test instructions that prove my fix is effective or that my feature works.

### CODE REVIEW

**Comment by pbiron** [comment]
Once this is merged, then all current uses of `esc_attr()` when adding text context to a sitemap (or index) should be replaced with `esc_xml()` and uses of `__()` for adding text to sitemaps/indexs/stylesheets should be replaced with `esc_xml__()`.

File: inc/functions.php [hunk 1]
```diff
@@ -117,3 +117,51 @@ function core_sitemaps_get_max_urls( $object_type ) {
...
+ */
+function esc_xml( $text ) { // phpcs:ignore WordPress.NamingConventions.PrefixAllGlobals
+	$safe_text = wp_check_invalid_utf8( $text );
+	$safe_text = _wp_specialchars( $safe_text, ENT_QUOTES );
+	$safe_text = html_entity_decode( $safe_text, ENT_HTML5 );
+	/**
+	 * Filters a string cleaned and escaped for output in XML.
+	 *
+	 * Text passed to esc_xml() is stripped of invalid or special characters
+	 * before output.  HTML named character references are converted to their
```
Comment (swissspidy) [inline inc/functions.php]:
```suggestion
	 * before output. HTML named character references are converted to their
```

File: inc/functions.php [hunk 2]
```diff
@@ -117,3 +117,51 @@ function core_sitemaps_get_max_urls( $object_type ) {
...
+	 * equiablent code points.
```
Comment (swissspidy) [inline inc/functions.php]:
```suggestion
	 * equivalent code points.
```

Comment (pbiron) [reply inc/functions.php]: oops...

**Comment by pbiron** [comment]
In core there is also `esc_html_e()` and `esc_html_x()`.  We don't need XML equivalents of those for sitemaps, but for the core merge they would be good.  @swissspidy  Should I add them to this PR?

**Comment by swissspidy** [comment]
@pbiron For completeness sake, why not 👍

btw, please let me know if the tests I added make sense.

**Comment by pbiron** [comment]
> @pbiron For completeness sake, why not 👍

will do.

> btw, please let me know if the tests I added make sense.

at first glance they look good.  will check in more detail shortly.

**Comment by swissspidy** [comment]
I presume we can also use these new functions already in the existing code base?

**Comment by pbiron** [comment]
> I presume we can also use these new functions already in the existing code base?

Yes, that's what I meant in https://github.com/GoogleChromeLabs/wp-sitemaps/pull/192#issuecomment-634086286

**Comment by pbiron** [comment]
thanx for the unit tests.  They pointed out that calling `html_entity_decode( $safe_text, ENT_HTML5 );` on the entire text is "too aggressive" (i.e., decodes too much, e.g., replaces `&amp;` with `&`).  I think I've got a fix for that, just trying to make it a little more efficient.

Also, `esc_xml()` should not do **any** modifications within [CDATA Sections](https://www.w3.org/TR/REC-xml/#dt-cdsection).  I'll work on that as soon as I make the replacement of HTML named character entities less aggressive.

**Comment by swissspidy** [comment]
> Also, `esc_xml()` should not do any modifications within CDATA Sections

How would you detect CDATA Sections? I'd say if someone uses them, they shouldn't be calling the function? We could suggest people not to do `'<![CDATA[' . esc_xml(...) . ']]>'`

**Comment by pbiron** [comment]
> > Also, `esc_xml()` should not do any modifications within CDATA Sections
> 
> How would you detect CDATA Sections? I'd say if someone uses them, they shouldn't be calling the function? We could suggest people not to do `'<![CDATA[' . esc_xml(...) . ']]>'`

that would be fine, if that's what they wanted.

The case I'm thinking of is plugin X does

```
update_post_meta(
   $post->ID,
   'my_meta_key',
   'This is a <![CDATA[test of the <emergency>]]> broadcast system'
);
```

and plugin Y (which adds custom properties to a sitemap, but doesn't have control over what plugin X wrote to post meta) does:
```
$url['some_custom_property'] = esc_xml( get_post_meta( $post->ID, 'my_meta_key', true ) );
```

**Comment by swissspidy** [comment]
Ah yes, makes sense 👍 Hope it's not too tricky to implement this, otherwise we go down a rabbit hole trying not to cause security issues...

**Comment by pbiron** [comment]
> Ah yes, makes sense 👍 Hope it's not too tricky to implement this, otherwise we go down a rabbit hole trying not to cause security issues...

Not hard, but it did take [ALOT](http://hyperboleandahalf.blogspot.com/2010/04/alot-is-better-than-you-at-everything.html) of experimentation to get the regex that separates out CDATA Sections both: 1) correct and 2) understandable by mere mortals :-)

File: inc/functions.php [hunk 3]
```diff
@@ -117,3 +117,146 @@ function wp_sitemaps_get_max_urls( $object_type ) {
...
+if ( ! function_exists( 'esc_xml_non_cdata_section' ) ) :
+	/**
+	 * Escaping for non-CDATA Section XML blocks.
+	 *
+	 * @since 5.5.0
+	 *
+	 * @param string $text Text to escape.
+	 * @return string
+	 */
+	function esc_xml_non_cdata_section( $text ) { // phpcs:ignore WordPress.NamingConventions.PrefixAllGlobals
```
Comment (swissspidy) [inline inc/functions.php]: IMO this should be renamed to `_esc_xml_non_cdata_section` and marked with `@access private`,  as its intended for internal use only.

**Review by swissspidy — APPROVED** [review]
LGTM with one nit

File: inc/functions.php (hunk 3 above)
Comment (pbiron) [reply inc/functions.php]: sounds good.

Will update shortly.

**Pull request merged**